eclair-core/src/main/scala/fr/acinq/eclair/db/psql/PsqlUtils.scala

```diff
@@ -0,0 +1,232 @@
+    val res = statement.executeQuery(s"SELECT version FROM versions WHERE db_name='$db_name'")
+    res.next()
+    res.getInt("version")
+  }
+
+  /**
+    * Updates the version for a particular logical database, it will overwrite the previous version.
+    */
+  def setVersion(statement: Statement, db_name: String, newVersion: Int): Unit = {
+    statement.executeUpdate("CREATE TABLE IF NOT EXISTS versions (db_name TEXT NOT NULL PRIMARY KEY, version INTEGER NOT NULL)")
+    // overwrite the existing version
+    statement.executeUpdate(s"UPDATE versions SET version=$newVersion WHERE db_name='$db_name'")
+  }
+
+  private def obtainDatabaseOwnership(instanceId: String, leaseDuration: FiniteDuration, attempt: Int = 1)(implicit ds: DataSource): Unit = synchronized {
+    logger.debug(s"Trying to acquire database ownership (attempt #$attempt) instance ID=${instanceId}")
+
+    if (attempt > 3) throw new TooManyLockAttempts("Too many attempts to acquire database ownership")
+
+    try {
+      inTransaction { implicit connection =>
+        acquireExclusiveTableLock()
+        getCurrentLease match {
+          case Some(lease) =>
+            if (lease.instanceId == instanceId || lease.expired)
+              updateLease(instanceId, leaseDuration)
+            else
+              throw new OwnershipException(s"The database is locked by instance ID=${lease.instanceId}")
+          case None =>
+            updateLease(instanceId, leaseDuration, insertNew = true)
+        }
+      }
+      logger.debug("Database ownership was successfully acquired.")
+    } catch {
+      case e: PSQLException if (e.getServerErrorMessage != null && e.getServerErrorMessage.getSQLState == "42P01") =>
+        withConnection {
+          connection =>
+            logger.warn(s"Table $OwnershipTable does not exist, trying to recreate it...")
+            initializeOwnershipTable(connection)
+            obtainDatabaseOwnership(instanceId, leaseDuration, attempt + 1)
+        }
+    }
+  }
+
+  private def initializeOwnershipTable(implicit connection: Connection): Unit = {
+    using(connection.createStatement()) {
+      statement =>
+        // allow only one row in the ownership lease table
+        statement.executeUpdate(s"CREATE TABLE IF NOT EXISTS $OwnershipTable (id INTEGER PRIMARY KEY default(1), expires_at TIMESTAMP NOT NULL, instance VARCHAR NOT NULL, CONSTRAINT one_row CHECK (id = 1))")
+    }
+  }
+
+  private def acquireExclusiveTableLock()(implicit connection: Connection): Unit = {
+    using(connection.createStatement()) {
+      statement =>
+        statement.executeUpdate(s"SET lock_timeout TO '${lockTimeout.toSeconds}s'")
+        statement.executeUpdate(s"LOCK TABLE $OwnershipTable IN ACCESS EXCLUSIVE MODE")
+    }
+  }
+
+  private def checkDatabaseOwnership(connection: Connection, instanceId: String, lockExceptionHandler: LockExceptionHandler): Unit = {
```

Comment:
Can you add a comment somewhere explaining how the locking works. It isn't immediately obvious. Also, is this a standard technique, or completely tailored made to our use case?

Reply:
It's a pretty much standard concept, but the implementation is tailored for this particular use case.